### PR TITLE
Handle platforms where pthread_t is a structure

### PR DESCRIPTION
**Pull request by Orvid**
These platforms include MSVC, and I don't know of any others. This implements the changes needed for this to work properly on MSVC.

### CODE REVIEW

File: folly/io/async/NotificationQueue.h [hunk 1]
```diff
@@ -225,7 +225,7 @@ class NotificationQueue {
     : eventfd_(-1),
       pipeFds_{-1, -1},
       advisoryMaxQueueSize_(maxSize),
-      pid_(getpid()),
+      pid_((pid_t)getpid()),
```
Comment (ariccio) [inline folly/io/async/NotificationQueue.h]: Huh?

Comment (Orvid) [inline folly/io/async/NotificationQueue.h]: With the windows implementation of pthread, it defines `pid_t` as a `void*`, and `getpid()` returns an `int`. MSVC doesn't like this implicit construction of a pointer from a raw integer and produces an error. Explicitly casting it solves the problem.

Comment (ariccio) [inline folly/io/async/NotificationQueue.h]: Ahh, ok. Any reason why it's not a C++ cast?

Comment (Orvid) [inline folly/io/async/NotificationQueue.h]: Personal habit.

File: folly/Portability.h [hunk 2]
```diff
@@ -284,6 +284,66 @@ inline size_t malloc_usable_size(void* ptr) {
...
+
+inline bool operator !=(pthread_t ptA, pthread_t ptB) {
+  return pthread_equal(ptA, ptB) == 0;
+}
+
+inline bool operator !(pthread_t ptA) {
+  return ptA == 0;
+}
+
+#define pthread_t_init (pthread_t{NULL, 0})
```
Comment (sgolemon) [inline folly/Portability.h]: Note from @yfeldblum
```
constexpr. Maybe name this pthread_zero.

(IIRC the _t stands for typedef, i.e., typedef struct pthread pthread_t;, and probably shouldn't be present in value names.)
```

Note from @meyering:
```
I too prefer pthread_zero or even pthread_0 -- sounds more like a value than pthread_t_init, which might evoke something that performs an initialization. That's a little closer to the uintmax_t -> UINTMAX_MAX mapping.
```

File: folly/Portability.h [hunk 3]
```diff
@@ -284,6 +284,66 @@ inline size_t malloc_usable_size(void* ptr) {
...
+
+inline int pthread_attr_getstack(
+  pthread_attr_t* attr,
+  void** stackaddr,
+  size_t* stacksize) {
+  pthread_attr_getstackaddr(attr, stackaddr);
+  pthread_attr_getstacksize(attr, stacksize);
```
Comment (sgolemon) [inline folly/Portability.h]: Note from @meyering:
```
Don't we want to return nonzero if either of these functions fails?
```

File: folly/Portability.h [hunk 4]
```diff
@@ -284,6 +284,66 @@ inline size_t malloc_usable_size(void* ptr) {
...
+  pthread_attr_setstacksize(attr, stacksize);
+  return 0;
+}
+
+inline int pthread_attr_getguardsize(pthread_attr_t* attr, size_t* guardsize) {
+  *guardsize = 0;
+  return 0;
+}
+#else
+#define pthread_t_init ((pthread_t)0)
```
Comment (sgolemon) [inline folly/Portability.h]: Note from @yfeldblum: `same`

File: folly/Portability.h (hunk 2 above)
Comment (Orvid) [inline folly/Portability.h]: I'll go with pthread_zero then. I believe I have an additional bit locally for this, and I need to implement a hasher for it for std::hash, as HHVM uses pthread_t as a key.

File: folly/Hash.h [hunk 5]
```diff
@@ -88,6 +92,14 @@ size_t hash_combine_generic(const T& t, const Ts&... ts) {
 // strings or pairs
 class StdHasher {
  public:
+#ifdef _MSC_VER
+  // std::hash doesn't support structs, but, on windows,
+  // pthread_t is a struct, so combine the hash of the
+  // two members.
+  static size_t hash(const pthread_t& t) {
```
Comment (Orvid) [inline folly/Hash.h]: This is what I need to relocate for proper hashing support.

**Comment by Orvid** [comment]
Alright, this has been updated to my local version, and also has eliminated the need for the hasher in Hash.h, and renamed pthread_t_init to pthread_zero.

**Comment by jwatzman** [comment]
@sgolemon is there an internal diff for this, sounds like it? What's the diff number, and what's the status?

**Comment by Orvid** [comment]
The publicly imported diff is [here](https://reviews.facebook.net/D46365) I don't know the internal diff number though.

**Comment by jwatzman** [comment]
Yeah, I eventually found it. I didn't realize we had our fancy PR process turned on for folly yet :)

**Comment by Orvid** [comment]
Well, last I checked it wasn't quite on fully, as it wasn't adding the links to the Phabricator reviews to new PRs...

**Comment by Orvid** [comment]
Replacing internally with D2862671.

**Pull request closed without merging**